src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,135 @@
+        private static readonly LocalizableString s_localizableMessage =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyMessage), Resx.ResourceManager, typeof(Resx));
+        private static readonly LocalizableString s_localizableDescription =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyDescription), Resx.ResourceManager, typeof(Resx));
+
+        public const string AdditionalDocumentLocationInfoSeparator = ";;";
+
+        public const string ConditionalOperation = nameof(ConditionalOperation);
+        public const string ChildStatementOperation = nameof(ChildStatementOperation);
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false,
+            additionalCustomTags: WellKnownDiagnosticTags.Unnecessary);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericDictionary2, out var dictionaryType))
+                return;
+
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.Conditional);
+
+            static void AnalyzeOperation(OperationAnalysisContext context)
+            {
+                var conditionalOperation = (IConditionalOperation)context.Operation;
+
+                IInvocationOperation? invocationOperation = null;
+
+                switch (conditionalOperation.Condition)
+                {
+                    case IInvocationOperation:
+                        invocationOperation = (IInvocationOperation)conditionalOperation.Condition;
+                        break;
+                    case IUnaryOperation unaryOperation when unaryOperation.OperatorKind == UnaryOperatorKind.Not:
+                        if (unaryOperation.Operand is IInvocationOperation operand)
+                            invocationOperation = operand;
+                        break;
+                    default:
+                        return;
+                }
+
+                if (invocationOperation!.TargetMethod.Name != "ContainsKey")
```

Comment:
This doesn't feel safe. I think `invocationOperation` can be null here.

Reply:
Yup, let's go with `if (invocationOperation == null || invocationOperation.TargetMethod.Name != "ContainsKey")` instead.

However, I can't really think of a case where it would be null, so I haven't added a test for this.

Reply:
@chucker I think it should be null for something like `if (variable)` or `if (expr1 && expr2)`. This code path will be executed for all conditional operations in the compilation.

Reply:
It would return in those cases, but you're right. If there's a `!` operator  _and_ no invocation operand, this did indeed throw. Added a test.